### PR TITLE
Add keycloak and oidc

### PR DESCRIPTION
**Pull request by soerface**
- Add keycloak with a preconfigured realm (contains banananick, email, drinks_notification setting)
- Add python class which communicates with keycloak to list users via a service account

TODO in later PRs:
- Move the "last drink notification" and "last emailed" fields from ldap to a local database
- Replace ldap list by keycloak user list

### CODE REVIEW

File: compose.dev.yaml
Comment (dargmuesli) [inline compose.dev.yaml:44]: This should be read form a `.env` file, but I guess that's already on your radar :wink:

File: drinks_touch/config.py
Comment (dargmuesli) [inline drinks_touch/config.py:14]: Reading a secret value in a PR is always emotionally challenging xD

File: drinks_touch/game.py
Comment (dargmuesli) [inline drinks_touch/game.py:109]: Can be removed? Or still useful for some testing?

File: package.json
Comment (dargmuesli) [inline package.json:17]: no conventional commits anymore? 🫨

File: compose.dev.yaml
Comment (soerface) [reply compose.dev.yaml:44]: Keycloak in compose is only needed for dev setup. I see no need to configure it via env vars

File: drinks_touch/config.py
Comment (soerface) [reply drinks_touch/config.py:14]: Sure, but the secret just applies to the local keycloak, so I don't see a risk, while at the same time it makes dev setup way more comfortable. Any suggestion for another solution to achieve this while mitigating some risk?

File: drinks_touch/game.py
Comment (soerface) [reply drinks_touch/game.py:109]: Could serve as a reminder on how to use the class, but we could certainly move it somewhere else. Maybe into the oidc.py file? Next steps would anyway be to move loading the users directly from ldap to instead load from the database, so right now this code isn't too relevant. I'll move it. Any other suggestion where to put it so we can find it later?

File: package.json
Comment (soerface) [reply package.json:17]: When I tried to commit, I always got some nodejs traceback. It didn't told me what I did wrong, just crashed. Honestly, I didn't see much use in those hooks for this particular project. Since I was anyway forced to skip the commit hooks, I decided to deactivate the automated installer so others don't run into the same problem.

Maybe I'm just ignorant and simply don't get the point. Would you mind explaining what we achieve with husky or conventional commits in this project?

**Comment by soerface** [comment]
@dargmuesli checkout my last commit. I also added an explanation about our intention, does this make sense?

What I'm also interested in, did this setup work for you? So, can you start the keycloak instance in the docker setup, interact with it, and does the example code work for you to list users?

File: compose.dev.yaml
Comment (dargmuesli) [reply compose.dev.yaml:44]: Ah, I see. Is there another Keycloak configured for production already? Didn't know of any yet.

File: package.json
Comment (dargmuesli) [reply package.json:17]: Using [conventional commits](https://www.conventionalcommits.org) is used to automatically create [releases for this project](https://github.com/flipdot/drinks-touch/releases) and to fill in the [changelog](https://github.com/flipdot/drinks-touch/blob/master/CHANGELOG.md). Right now those contain the weekly automatic release only as there has been little development going on. So what spending a thought on the commit title lets you benefit is that if you implement a `feat: new component` the project (versioned at 0.4.47) would automatically release a version 0.5.0. Or if you implement a `fix: correct variable name` version 0.4.48 would be released, all without you having to manually go through some release process.

See some insightful example where more frequent development is going on over at: [dargmuesli/nuxt-cookie-control/releases](https://github.com/dargmuesli/nuxt-cookie-control/releases).

It's just nice to never have to think about releases again, just by typing following a conventional format for commit messages.

The weekly automated releases are just some basic strategy to *release* dependency updates that come in by renovate from time to time. If there was some security fix in some dependency it would not take too long to be made available for deployment (and in the perfect world be covered by tests and even deployed for you). Necessity for this project is arguable due to the non-public and manual deployment, so we can remove that to not spam the changelog or turn the frequency down. I just had set it up as I planned to develop more on the "perfect world" approach, for which I haven't found time yet apparently :grin:

**Comment by dargmuesli** [comment]
I haven't booted the project yet and am currently out of time to do so, I just looked over the code here on GitHub :raised_hands:

File: compose.dev.yaml
Comment (soerface) [reply compose.dev.yaml:44]: Yes, checkout https://login.flipdot.org/, you can login with your existing flipdot LDAP credentials :) Some of our applications already support a login with it, checkout the "Applications" section in keycloak after login (door not working yet, the rest of the applications do, but some need manual linking or migration effort beforehand). This now allows you to use passkeys to login at various flipdot services!

Comment (dargmuesli) [reply compose.dev.yaml:44]: Wow, nice! :fire: I smell some incoming talks between us about this architecture in October :wink:

File: keycloak/realms/flipdot.json
Comment (malled2002) [inline keycloak/realms/flipdot.json:749]: Script upload is disabled. So this section prevents the import form the flipdot realm.
https://stackoverflow.com/questions/72128765/import-realm-in-keycloak18-x

File: drinks_touch/game.py
Comment (malled2002) [inline drinks_touch/game.py:12]: is this necessary?

**Review by malled2002 — COMMENTED** [review]
is this realm import error just on my setup?
can we fix this without manually edit the exported file?

File: drinks_touch/game.py
Comment (soerface) [reply drinks_touch/game.py:12]: Nope, good catch. Also a couple other unused imports here, will remove them.

How do you feel about adding black & flake8 with pre-commit so things like this get automatically solved?

File: keycloak/realms/flipdot.json
Comment (soerface) [reply keycloak/realms/flipdot.json]: This is super weird. So the whole section needs to go? I was able to fix it by dropping it from this file, but everytime you export a realm, it get's added again by keycloak. Following your link there can also be found this discussion, where they talk about keycloak exporting stuff that isn't importable whenever a confidential client is involved: https://github.com/keycloak/keycloak/discussions/12041#discussioncomment-2768768

I'm not even sure what the section is good for. At least the password is still part of the json, so the example code still works with this change:

https://github.com/flipdot/drinks-touch/pull/213/commits/c1d6d9268704a9adbe060e080f7c5e1d42eccc34

**Comment by soerface** [comment]
@malled2002 fixed the import issue and removed unused imports. Squashed commits and added sensible commit messages. Ready for second review.

**Review by malled2002 — APPROVED** [review]
looks good to me :+1:

File: .github/workflows/ci.yml
Comment (soerface) [reply .github/workflows/ci.yml:48]: @dargmuesli Can we make this job to run only on PRs?

Comment (dargmuesli) [reply .github/workflows/ci.yml:48]: Another approach (with arguable performance improvement) would be to fetch one commit initially and then to run `git fetch --deepen=1` (or 10, ...) until `git rev-list --count $(git merge-base HEAD origin/master)..HEAD` works aka the divergence from `master` can be found using git.

Comment (dargmuesli) [reply .github/workflows/ci.yml:48]: See #219

**Pull request merged**
